Storybook setup, Stories & unit tests for initial set of components

pavish opened the pull request:
Fixes #297 
Fixes #288 

### This PR:
* Sets up Storybook for component design system
* Adds stories for Button, TextInput, Icon and Skeleton components
* Adds unit tests for the same components
* Updates npm packages to their respective latest versions

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

kgodey commented on mathesar_ui/README.md:
```diff
@@ -63,6 +63,22 @@ node@c273da65c52d:/ui$ npm install <package>
 
+### Components
+
+For guidelines on component development, refer [README of components](https://github.com/centerofci/mathesar/tree/integrate_storybook/mathesar_ui/src/components#readme).
```
Also, I don't see this file in this PR.

pavish replied:
My bad. I've modified it to `master`. The file is already present in current master, and it's not part of this PR.

kgodey commented on mathesar_ui/README.md:
````diff
@@ -63,6 +63,22 @@ node@c273da65c52d:/ui$ npm install <package>
+```bash
+docker exec -it mathesar_ui_1 /bin/bash
+
+node@c273da65c52d:/ui$ npm run storybook
````
These instructions are not working for me.

```
node@62013923826d:~/ui$ npm run storybook

> mathesar_ui@0.0.0 storybook /home/node/ui
> start-storybook -p 6006

sh: 1: start-storybook: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! mathesar_ui@0.0.0 storybook: `start-storybook -p 6006`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the mathesar_ui@0.0.0 storybook script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2021-06-30T19_58_17_245Z-debug.log
```

dhruvkb replied:
This could be an issue with the Storybook bins not present in `node_modules/`. Try by rebuilding the container?

pavish replied:
This issue is because storybook was not installed. Restarting the container would perform an `npm install`, and should install the required packages.

kgodey reviewed (changes requested):
@pavish The code looks good but I haven't been able to run the storybook locally.

@dhruvkb or @zackkrida, it would be great to get a review on this since you're more familiar with Storybook.

dhruvkb commented:
I can't get the Storybook to run either. SCSS imports are not working. The log below shows 'Skeleton.scss' but it's a different component every time I run `npm run storybook`.

```
WARN Broken build, fix the error above.
WARN You may need to refresh the browser.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! mathesar_ui@0.0.0 storybook: `start-storybook -p 6006`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the mathesar_ui@0.0.0 storybook script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2021-06-30T20_14_06_294Z-debug.log
node@3c88ad2042a2:~/ui$ Error: Can't find stylesheet to import.
>   ╷
> 2 │   @import "Skeleton.scss";
>   │           ^^^^^^^^^^^^^^^
```

zackkrida reviewed (changes requested):
I expected to be able to `docker compose up --build` but I wasn't able to see anything running at the `3000` or `6001` ports as expected from the ui container.

Same issue as @Dhruvi16 when I try to run or build storybook in the `mathesar_ui` repo. I imagine this is related to preprocessing the `scss` styles within Svelte.

pavish commented:
@zackkrida The port `3000` does not serve any html files. It runs the vite js server, and serves the `<host>:3000/@vite/client` and `<host>:3000/src/main.ts` (i.e. all source files). Checking `3000` directly would not show anything.

These files are referenced in our django templates, and when the web server runs in `8000`, the client files are loaded from the vite server running in `3000`.

We do not use `6001`. Storybook uses `6006`, and when storybook is running in dev mode, `6006` would show the storybook ui.

I will check what is causing the issues with scss import. It seems to be running fine for me, so it might take a while to reproduce the issue and find the root cause.

pavish commented:
I've added a commit 'Running scss preprocess synchronously for storybook', which is a fix for the issue described by @dhruvkb and @zackkrida .

## RCA:

## Reproducing the issue:
I wasn't initally able to reproduce this issue. I tried running storybook in dev mode and storybook build multiple times, both in the container and directly, and was able to reproduce the issue once, when my CPU consumption was high.

I had assigned 6 CPUs for my Docker. I reduced it to 2, and was able to reproduce the issue everytime I ran storybook.

## Identifying the cause:
The error message wasn't particularly helpful. 

```
ModuleBuildError: Module build failed (from ./node_modules/@storybook/addon-docs/dist/cjs/frameworks/svelte/svelte-docgen-loader.js):
Error: Can't find stylesheet to import.
  ╷
2 │   @import "TextInput.scss";
  │           ^^^^^^^^^^^^^^^^
  ╵
```

I tried removing the stories for other components, and only ran with TextInput, and _it succeeded_. Weird!

When I ran with stories for more than 2 components, the issue reproduced. This lead to the realization that this is happening when there are multiple stylesheets being processed.

But where the issue occurred still eluded me, is it in the Storybook docgen loader, or in svelte-preprocessor?

I snooped around the docgen loader code, and found that they only call the preprocessor, and do nothing additionally. If the issue was in the preprocessor, then we should have seen it while running our vite dev server or build.

So, neither of these were causing the issue.

There was another clue in the logs.

```
  src/components/icon/Icon.svelte 2:11  root stylesheet
    at Object._newRenderError (/home/node/ui/node_modules/sass/sass.dart.js:13222:19)
    at Object._wrapException (/home/node/ui/node_modules/sass/sass.dart.js:13050:16)
    ...
    at /home/node/ui/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /home/node/ui/node_modules/loader-runner/lib/LoaderRunner.js:233:18
```

The failed import was on TextInput.scss, but the error here states that Icon.svelte threw an error. It didn't make sense.
But as clearly noticable, this was thrown by **dart-sass**. But why?

Remembering that this only happens when multiple files are being processed, I realized it could be a race condition when it is all happening in parallel.

I tried running scss preprocessing in sync. The error did not reproduce. I tried running it several times again, with low resources, but the issue didn't reproduce with this configuration. Voila! We have the possible cause. 

### The fix
In the commit, I've added a separate preprocess configuration where scss is processed synchronously. I've added this only for storybook, not for vite. Based on sass docs, [they've asked us to prefer renderSync over render](https://sass-lang.com/documentation/js-api#render).

### Next steps
If infact dart-sass is causing the issue underneath, we might see similar issues later on with our vite server as well, in which case we would have to update the config for vite. Nevertheless, we need to keep an eye out for this issue in the future.

For now, we have a working configuration. I'm not very sure where to raise this issue though (svelte? storybook? or sass? Even though sass is responsible for the actual error, maybe it depends on how svelte or storybook configures it internally).

zackkrida reviewed (approved):
Can confirm storybook now works locally, after doing the following:

```shell
git pull
cd mathesar_ui
rm -rf node_modules
npm install
npm run build
npm run storybook
```

I did have to remove `node_modules/` for things to work, simply pulling and starting storybook led to a new 500 error in the browser (even though the storybook process ran successfully)

kgodey reviewed (approved):
Looks good to me! Storybook runs locally for me now, and I like the level of detail 👍 .

Thanks also for the detailed root cause analysis, @pavish.

kgodey commented:
I'm going to merge. @dhruvkb feel free to review and we can address any comments if you have them in a future PR.
